src/debug_mode.jl

```diff
@@ -1,5 +1,78 @@
-# TODO: make it non-trivial. See https://github.com/chalk-lab/Mooncake.jl/issues/672
-DebugFRule(rule) = rule
+"""
+    DebugFRule(rule)
+
+Construct a callable equivalent to `rule` but with additional type checking for forward-mode
+AD. Checks:
+- Each `Dual` argument has tangent type matching `tangent_type(typeof(primal))`
+- The returned `Dual` has a correctly-typed tangent
+- Deep structural validation (array sizes, field types, etc.)
+
+Forward-mode counterpart to [`DebugRRule`](@ref).
+
+*Note:* Debug mode significantly slows execution (10-100x) and should only be used for
+diagnosing problems, not production runs.
+```
+"""
+struct DebugFRule{Trule}
+    rule::Trule
+end
+
+# Recursively copy the wrapped rule
+_copy(x::P) where {P<:DebugFRule} = P(_copy(x.rule))
+
+"""
+    (rule::DebugFRule)(x::Vararg{Dual,N}) where {N}
+
+Apply pre- and post-condition type checking. See [`DebugFRule`](@ref).
+"""
+@noinline function (rule::DebugFRule)(x::Vararg{Dual,N}) where {N}
+    verify_dual_inputs(rule.rule, x)
+    y = rule.rule(x...)
+    verify_dual_output(x, y)
+    return y::Dual
+end
+
+@noinline function verify_dual_inputs(rule, @nospecialize(x::Tuple))
+    try
+        for _x in x
+            _x isa Dual || error("Expected Dual, got $(typeof(_x))")
+            verify_dual_value(_x)
+        end
+    catch e
+        error("error in inputs to rule with input types $(_typeof(x))")
+    end
+end
+
+@noinline function verify_dual_output(@nospecialize(x), @nospecialize(y))
+    try
+        y isa Dual || error("frule!! must return a Dual, got $(typeof(y))")
+        verify_dual_value(y)
+    catch e
+        error("error in outputs of rule with input types $(_typeof(x))")
+    end
+end
+
+@noinline function verify_dual_value(d::Dual)
+    p, t = primal(d), tangent(d)
+
+    # Fast path: type-level check
+    T_expected = tangent_type(typeof(p))
+    T_actual = typeof(t)
+    if !(T_actual <: T_expected)
+        throw(
+            InvalidFDataException(
+                "Dual tangent type mismatch: primal $(typeof(p)) requires tangent type " *
+                "$T_expected, but got $T_actual",
+            ),
+        )
+    end
+
+    # Slow path: deep structural validation
```

Comment:
Do you want to add a comment that we check `fdata` and `rdata` separately such that the logic can be shared with reverse mode?